hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/DefaultCompactWriter.java

```diff
@@ -425,12 +421,16 @@ protected FieldDescriptor checkFieldDefinition(@Nonnull String fieldName, @Nonnu
     @Override
     public <T> void writeArrayOfCompact(@Nonnull String fieldName, @Nullable T[] value) {
         writeArrayOfVariableSize(fieldName, ARRAY_OF_COMPACT, value,
-                (out, val) -> serializer.writeObject(out, val, includeSchemaOnBinary));
+                new SingleTypeCompactArrayItemWriter<>(
```

Comment:
Will this increase litter? Creating one more object compared to the old impl. Can you eliminate this extra object creation? Would it be too much cost to check the type of the element in previous index and if it is the first element no checks at all?

Reply:
Made it allocate one less object in https://github.com/hazelcast/hazelcast/pull/21958/commits/7f2319a57c3e2a9638975b3ae68ed10fd1c7ef56

Regarding the other question, to do that, I would have to change the `writeArrayOfVariableSize` which is used in a lot of places. I don't think it is such a crucial to perform one less if check so I am voting to leave things as it is.

Reply:
We need to study on litter overall for the whole compact implementation to make some places more efficient but we said that this can be done later on as an optimization.